clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,142 @@
+= Distributed EJB timers
+:author:            Paul Ferraro
+:email:             paul.m.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, WildFly supports 2 mechanisms for persistent EJB timers: a file-based implementation and a JDBC-based implementation.
+Of these, only the JDBC-based implementation is suitable for use in a cluster.   However, this implementation does not scale well to large to medium sized clusters and relies on a single point of failure.
+WildFly needs an persistent timer service implementation suitable for use in a cluster is both highly-available and reasonably scalable.
+
```

Comment:
What would mean for an end user to adopt this feature? Both from the developer and the administration perspective? e.g.: how is existing code/developer experience affected? How would an admin move from a current JDBC based solution to the new one?

Of course this cannot be detailed here, but a high level description, examples code or commands would definitely help the reader to have a clear vision of the scope of what would mean to migrate to this feature.

Also, it would be definitely important to have some, even very short and simple, consideration - literally from you - about how this feature could affect perfomance, since there are some issues in the distributed EJB timers portfolio around which happen(ed) to pop up in such context (e.g. high load). This would be of inspiration for more complex testing scenarios.

Reply:
> What would mean for an end user to adopt this feature? Both from the developer and the administration perspective? e.g.: how is existing code/developer experience affected? How would an admin move from a current JDBC based solution to the new one?

The developer experience is identical.
The administrator experience is simpler, since there would no longer be the need to provision and configure a database resource.

> Of course this cannot be detailed here, but a high level description, examples code or commands would definitely help the reader to have a clear vision of the scope of what would mean to migrate to this feature.

This is included in the community documentation.

> Also, it would be definitely important to have some, even very short and simple, consideration - literally from you - about how this feature could affect performance,

The root cause of the performance issues with the existing database timer-store implementation is that it relies on blocking mechanisms. Timer "ownership" is not deterministic, but rather is determined by which EAP instance is able to obtain an exclusive lock on a given table entry.  While the instance that is able to lock the row will return immediately, due to the nature of the JDBC API and relational database locking in general, instances contending for the same row lock will block.  When an instance eventually does obtain the lock, the associated timer will have already scheduled it, and thus each instance will try to lock the next timer, and so on.  For obvious reasons, this does not scale.

We solve this in the new implementation be making timer ownership deterministic, i.e. an intrinsic function of the cluster topology and the timer identifier; in the same manner that EAP manages all other distributed state.  Thus the new implementation does not rely on blocking mechanisms in order to schedule timers.

Memory footprint issues are mitigated through the use of Inifnispan cache passivation.

> since there are some issues in the distributed EJB timers portfolio around which happen(ed) to pop up in such context (e.g. poor performance if high throughput and memory footprint if high load). This would be of inspiration for more complex testing scenarios.

I anticipate that the same QE tests used to reproduce these performance issues can also be used to validate that the new implementation resolves them.

Reply:
> Thanks for sharing these details. I think they add great value and I am wondering if there's any reason for not including the above explanation into the document itself. WDYT?

The justification of this feature has already been discussed in the associated JIRAs.  The purpose of this document is to analyze the feature requirements.

Reply:
> Ok, thanks, then could you please mention this when you'll add the Community docs section contents?

Will do.